app/merginapi.cpp

```diff
@@ -1127,75 +1168,68 @@ void MerginApi::uploadInfoReplyFinished()
   Q_ASSERT( r );
 
-  QPair<QHash<QString, QList<MerginFile>>, QString> pair = parseAndCompareProjectFiles( r, false );
-  QHash<QString, QList<MerginFile>> files = pair.first;
-  QString version = pair.second;
-
   QUrl url = r->url();
   mPendingRequests.remove( url );
   QString projectNamespace;
   QString projectName;
   extractProjectName( url.path(), projectNamespace, projectName );
   QString projectFullName = getFullProjectName( projectNamespace, projectName );
 
-  if ( r->error() != QNetworkReply::NoError )
+  if ( r->error() == QNetworkReply::NoError )
   {
-    emit syncProjectFinished( QStringLiteral(), projectFullName, false );
-  }
+    QJsonObject changes;
+    InputUtils::log( r->url().toString(), QStringLiteral( "FINISHED" ) );
+    QByteArray data = r->readAll();
 
-  std::shared_ptr<MerginProject> project = getProject( projectFullName );
-  QString projectPath = getProjectDir( projectNamespace, projectName ) + "/";
+    QString projectPath = getProjectDir( projectNamespace, projectName ) + "/";
+    QList<MerginFile> localFiles = getLocalProjectFiles( projectPath );
 
-  QJsonDocument jsonDoc;
-  QJsonObject changes;
+    MerginProject serverProject = readProjectMetadata( data );
+    mTempMerginProjects.insert( projectNamespace + "/" + projectName, serverProject );
 
-  QList<MerginFile> filesToUpload;
-  for ( QString key : files.keys() )
-  {
+    ProjectDiff diff = compareProjectFiles( localFiles, serverProject.files );
+    QList<MerginFile> filesToUpload;
     QJsonArray jsonArray;
-    for ( MerginFile file : files.value( key ) )
-    {
-      QJsonObject fileObject;
-      fileObject.insert( "path", file.path );
-      fileObject.insert( "checksum", file.checksum );
-      fileObject.insert( "size", file.size );
-      fileObject.insert( "mtime", file.mtime.toString( Qt::ISODateWithMs ) );
 
-      qreal rawNoOfChunks = qreal( file.size ) / UPLOAD_CHUNK_SIZE;
-      int noOfChunks = qCeil( rawNoOfChunks );
-      QStringList chunks = generateChunkIds( noOfChunks );
-      QJsonArray chunksJson;
-      for ( QString id : chunks )
-      {
-        chunksJson.append( id );
-      }
-      file.chunks = chunks;
-      fileObject.insert( "chunks", chunksJson );
-      jsonArray.append( fileObject );
+    QJsonObject added = prepareUploadChangesJSON( diff.added );
+    jsonArray.append( added );
+    filesToUpload.append( diff.added );
 
-      if ( key != QStringLiteral( "removed" ) )
-      {
-        filesToUpload.append( file );
-      }
-    }
-    changes.insert( key, jsonArray );
-  }
+    QJsonObject modified = prepareUploadChangesJSON( diff.modified );
+    jsonArray.append( modified );
+    filesToUpload.append( diff.modified );
 
-  r->deleteLater();
-  mPendingRequests.remove( url );
-  mFilesToUpload.insert( projectFullName, filesToUpload );
+    QJsonObject removed = prepareUploadChangesJSON( diff.removed );
+    jsonArray.append( removed );
+    // removed not in filesToUpload
```

Comment:
Removed files probably should be passed to the server just as a list of file names, rather than having full file info?